### PR TITLE
Access key and secret key cant start with '-'.

### LINKED ISSUE #6684: Access key and Secret key cannot have "-" as first character
When minio generates a secret key, or if user uses environment variable and has "-" as first character, when you try to add the server it assumes the key is an argument

## Expected Behavior

We should not allow "-" to be the first character of the secret key or access key, and we should never automatically generate keys with "-" in the first position.

## Current Behavior

When trying to add via mc you see the message:
Incorrect Usage: flag provided but not defined: -minio

## Steps to Reproduce (for bugs)

1. Create minio instance with secret key or access key starting with "-" char
```
export MINIO_ACCESS_KEY="-minio"; export MINIO_SECRET_KEY="secretkey123"; minio server /data 
```
2. Try to add host via `mc`
```
mc config host add badkey http://127.0.0.1:9000 -minio secretkey123
Incorrect Usage: flag provided but not defined: -minio
```

## Context

Initially saw this issue with auto generated secret key, so we need to handle the case where user generated or server generated automatically
## Regression

Unknown
## Your Environment

* Version used (`minio version`):
Version: 2018-10-18T00:28:58Z
Release-Tag: RELEASE.2018-10-18T00-28-58Z
Commit-ID: 3b553570459cdf8b10671105252c707ff2cf76f3
* Operating System and version (`uname -a`):
Tested against Minio running in Mac OS High Sierra and Windows 2012 R2

### PR DESCRIPTION
**Pull request by aarushiarya**
Fixes #6684 
Access key and secret key can't have first character '-'.

## Description

1. Check first character of automatically generated secret key. Regenerate if first character is '-'.
2. Check access key and secret key, set as environment variable. Fail with appropriate error message.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [ ] All new and existing tests passed.

### CODE REVIEW

File: pkg/auth/credentials.go [hunk 1]
```diff
@@ -132,6 +145,17 @@ func GetNewCredentialsWithMetadata(m map[string]interface{}, tokenSecret string)
 		return cred, err
 	}
 	cred.SecretKey = string([]byte(base64.URLEncoding.EncodeToString(keyBytes))[:secretKeyMaxLen])
+	// Regenerate secret key until first character is not '-'
+	for {
+		if (cred.SecretKey)[0] != '-' {
```
Comment (harshavardhana) [inline pkg/auth/credentials.go]: "-" is never possible in base64 @aarushiarya to be the first character.

File: pkg/auth/credentials.go [hunk 2]
```diff
@@ -87,6 +91,15 @@ func (cred Credentials) IsExpired() bool {
 func (cred Credentials) IsValid() bool {
 	// Verify credentials if its enabled or not set.
 	if cred.Status == "enabled" || cred.Status == "" {
+		// Check if first character is '-'
+		if IsAccessKeyValid(cred.AccessKey) && (cred.AccessKey)[0] == '-' {
+			logger.FatalIf(ErrInvalidAccessKey, "Invalid access key")
+			os.Exit(1)
```
Comment (harshavardhana) [inline pkg/auth/credentials.go]: you cannot os.Exit(1) here its a package.

File: pkg/auth/credentials.go [hunk 3]
```diff
@@ -21,9 +21,11 @@ import (
 	"crypto/subtle"
 	"encoding/base64"
 	"fmt"
+	"os"
 	"time"
 
 	jwtgo "github.com/dgrijalva/jwt-go"
+	"github.com/minio/minio/cmd/logger"
```
Comment (harshavardhana) [inline pkg/auth/credentials.go]: Logger cannot be used in external packages its only meant to be used inside `cmd/`

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#fe0adc3f-484c-4189-a414-6b13d3368a28) |
| mint-gateway-nas.sh | :x: [more...](#f50f883a-0ef6-4970-b484-b032d0bc4642) |
| mint-large-bucket.sh | :x: [more...](#8d8d6b4b-e5e8-4d58-a0b3-3c60e0dc0c98) |
| mint-fs.sh | :x: [more...](#b55afaf5-f70c-497b-9773-f0236e6b7ea6) |
| mint-dist-xl.sh | :x: [more...](#7e76f529-db55-476d-a6ac-2b0ac20ed741) |
<a name="f50f883a-0ef6-4970-b484-b032d0bc4642"></a>
### 6731-784c9be/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30490
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0 
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 642a08207f43:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 25 seconds
{
  "name": "awscli",
  "duration": 9769,
  "function": "aws --endpoint-url http://minikube:30490 s3api head-object --bucket awscli-mint-test-bucket-16446 --key prefix/directory\n",
  "status": "FAIL",
  "error": "An error occurred (404) when calling the HeadObject operation: Not Found"
}

Executed 0 out of 13 tests successfully.
```
<a name="fe0adc3f-484c-4189-a414-6b13d3368a28"></a>
### 6731-784c9be/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31336
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0 
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 97513ee6c5e9:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 26 seconds
{
  "name": "awscli",
  "duration": 9856,
  "function": "aws --endpoint-url http://minikube:31336 s3api head-object --bucket awscli-mint-test-bucket-24600 --key prefix/directory\n",
  "status": "FAIL",
  "error": "An error occurred (404) when calling the HeadObject operation: Not Found"
}

Executed 0 out of 13 tests successfully.
```
<a name="7e76f529-db55-476d-a6ac-2b0ac20ed741"></a>
### 6731-784c9be/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:32220
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0 
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 82958ca0e971:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 12 seconds
{
  "name": "awscli",
  "duration": 4914,
  "function": "aws --endpoint-url http://minikube:32220 s3api head-object --bucket awscli-mint-test-bucket-13886 --key prefix/directory\n",
  "status": "FAIL",
  "error": "An error occurred (404) when calling the HeadObject operation: Not Found"
}

Executed 0 out of 13 tests successfully.
```
<a name="8d8d6b4b-e5e8-4d58-a0b3-3c60e0dc0c98"></a>
### 6731-784c9be/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30040
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0 
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 4f2020317b38:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 14 seconds
{
  "name": "awscli",
  "duration": 5613,
  "function": "aws --endpoint-url http://minikube:30040 s3api head-object --bucket awscli-mint-test-bucket-10903 --key prefix/directory\n",
  "status": "FAIL",
  "error": "An error occurred (404) when calling the HeadObject operation: Not Found"
}

Executed 0 out of 13 tests successfully.
```
<a name="b55afaf5-f70c-497b-9773-f0236e6b7ea6"></a>
### 6731-784c9be/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31393
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0 
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 7a76aaac9e0f:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 24 seconds
{
  "name": "awscli",
  "duration": 9732,
  "function": "aws --endpoint-url http://minikube:31393 s3api head-object --bucket awscli-mint-test-bucket-23179 --key prefix/directory\n",
  "status": "FAIL",
  "error": "An error occurred (404) when calling the HeadObject operation: Not Found"
}

Executed 0 out of 13 tests successfully.
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]
This PR is really not needed 

```
~ mc config host add --api s3v4 badkey https://127.0.0.1:9000 "-minio" "secretkey123"
```

This is actually a limitation of command line library providing `-` should be rejected by `mc` instead of server.

**Review by harshavardhana — CHANGES_REQUESTED** [review]
Actually discussed with @eco-minio since we are using URL encoding it is generating keys with `-` we should avoid that and instead use the following.

```diff
diff --git a/pkg/auth/credentials.go b/pkg/auth/credentials.go
index a9b8f1c93..5e4b5c867 100644
--- a/pkg/auth/credentials.go
+++ b/pkg/auth/credentials.go
@@ -21,6 +21,7 @@ import (
        "crypto/subtle"
        "encoding/base64"
        "fmt"
+       "strings"
        "time"
 
        jwtgo "github.com/dgrijalva/jwt-go"
@@ -131,7 +132,7 @@ func GetNewCredentialsWithMetadata(m map[string]interface{}, tokenSecret string)
        if err != nil {
                return cred, err
        }
-       cred.SecretKey = string([]byte(base64.URLEncoding.EncodeToString(keyBytes))[:secretKeyMaxLen])
+       cred.SecretKey = strings.Replace(string([]byte(base64.StdEncoding.EncodeToString(keyBytes))[:secretKeyMaxLen]), "/", "+", -1)
        cred.Status = "enabled"
 
        expiry, ok := m["exp"].(float64)
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#6114c58d-7b97-4c5c-ac93-a3da37f31773) |
| mint-gateway-nas.sh | :x: [more...](#3b90638d-bbff-4ab6-85fc-62fd4f43415f) |
| mint-large-bucket.sh | :x: [more...](#322d77d8-56a1-4051-a014-9894af6169e2) |
| mint-fs.sh | :x: [more...](#ef8c118a-5a62-48d3-9410-54f176a95f72) |
| mint-dist-xl.sh | :x: [more...](#251530f3-0acd-4939-8c1b-fa546484a658) |
<a name="322d77d8-56a1-4051-a014-9894af6169e2"></a>
### 6731-203762c/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30494
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0 
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp b41f26b35059:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 13 seconds
{
  "name": "awscli",
  "duration": 5923,
  "function": "aws --endpoint-url http://minikube:30494 s3api head-object --bucket awscli-mint-test-bucket-8062 --key prefix/directory\n",
  "status": "FAIL",
  "error": "An error occurred (404) when calling the HeadObject operation: Not Found"
}

Executed 0 out of 13 tests successfully.
```
<a name="3b90638d-bbff-4ab6-85fc-62fd4f43415f"></a>
### 6731-203762c/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:32486
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0 
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 2d4d5743fbf4:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 24 seconds
{
  "name": "awscli",
  "duration": 10290,
  "function": "aws --endpoint-url http://minikube:32486 s3api head-object --bucket awscli-mint-test-bucket-5522 --key prefix/directory\n",
  "status": "FAIL",
  "error": "An error occurred (404) when calling the HeadObject operation: Not Found"
}

Executed 0 out of 13 tests successfully.
```
<a name="251530f3-0acd-4939-8c1b-fa546484a658"></a>
### 6731-203762c/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30364
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0 
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 1b51d6d2a8e6:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 11 seconds
{
  "name": "awscli",
  "duration": 4555,
  "function": "aws --endpoint-url http://minikube:30364 s3api head-object --bucket awscli-mint-test-bucket-16130 --key prefix/directory\n",
  "status": "FAIL",
  "error": "An error occurred (404) when calling the HeadObject operation: Not Found"
}

Executed 0 out of 13 tests successfully.
```
<a name="6114c58d-7b97-4c5c-ac93-a3da37f31773"></a>
### 6731-203762c/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:32605
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0 
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 14c9fc8ef76c:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 24 seconds
{
  "name": "awscli",
  "duration": 9781,
  "function": "aws --endpoint-url http://minikube:32605 s3api head-object --bucket awscli-mint-test-bucket-26218 --key prefix/directory\n",
  "status": "FAIL",
  "error": "An error occurred (404) when calling the HeadObject operation: Not Found"
}

Executed 0 out of 13 tests successfully.
```
<a name="ef8c118a-5a62-48d3-9410-54f176a95f72"></a>
### 6731-203762c/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30066
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0 
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp e931ab9892a1:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 23 seconds
{
  "name": "awscli",
  "duration": 10143,
  "function": "aws --endpoint-url http://minikube:30066 s3api head-object --bucket awscli-mint-test-bucket-1763 --key prefix/directory\n",
  "status": "FAIL",
  "error": "An error occurred (404) when calling the HeadObject operation: Not Found"
}

Executed 0 out of 13 tests successfully.
```

**Review by harshavardhana — APPROVED** [review]
LGTM

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#8b6629cd-8e95-4b44-b132-2bebda93feba) |
| mint-gateway-nas.sh | :x: [more...](#7bf97568-81a2-45cb-a59e-dd0a16f0b82b) |
| mint-large-bucket.sh | :x: [more...](#13f13716-9a6e-4143-a594-7afce3f0755b) |
| mint-fs.sh | :x: [more...](#581acfd8-7bac-46cb-bc75-2c263af8370f) |
| mint-dist-xl.sh | :x: [more...](#62c7b39c-5e72-426a-858a-07f6ad830965) |
<a name="13f13716-9a6e-4143-a594-7afce3f0755b"></a>
### 6731-3ae0b37/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30727
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0 
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 109f4c1f164f:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 15 seconds
{
  "name": "awscli",
  "duration": 5600,
  "function": "aws --endpoint-url http://minikube:30727 s3api head-object --bucket awscli-mint-test-bucket-31722 --key prefix/directory\n",
  "status": "FAIL",
  "error": "An error occurred (404) when calling the HeadObject operation: Not Found"
}

Executed 0 out of 13 tests successfully.
```
<a name="8b6629cd-8e95-4b44-b132-2bebda93feba"></a>
### 6731-3ae0b37/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30183
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0 
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp fc96ceb7de3a:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 25 seconds
{
  "name": "awscli",
  "duration": 10176,
  "function": "aws --endpoint-url http://minikube:30183 s3api head-object --bucket awscli-mint-test-bucket-24529 --key prefix/directory\n",
  "status": "FAIL",
  "error": "An error occurred (404) when calling the HeadObject operation: Not Found"
}

Executed 0 out of 13 tests successfully.
```
<a name="7bf97568-81a2-45cb-a59e-dd0a16f0b82b"></a>
### 6731-3ae0b37/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:32289
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0 
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 1bd9fdc7660e:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 25 seconds
{
  "name": "awscli",
  "duration": 10110,
  "function": "aws --endpoint-url http://minikube:32289 s3api head-object --bucket awscli-mint-test-bucket-12141 --key prefix/directory\n",
  "status": "FAIL",
  "error": "An error occurred (404) when calling the HeadObject operation: Not Found"
}

Executed 0 out of 13 tests successfully.
```
<a name="62c7b39c-5e72-426a-858a-07f6ad830965"></a>
### 6731-3ae0b37/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31694
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0 
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp e8c7ddbe9bc4:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 10 seconds
{
  "name": "awscli",
  "duration": 4619,
  "function": "aws --endpoint-url http://minikube:31694 s3api head-object --bucket awscli-mint-test-bucket-9462 --key prefix/directory\n",
  "status": "FAIL",
  "error": "An error occurred (404) when calling the HeadObject operation: Not Found"
}

Executed 0 out of 13 tests successfully.
```
<a name="581acfd8-7bac-46cb-bc75-2c263af8370f"></a>
### 6731-3ae0b37/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30126
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0 
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 0bf8f05cd9b9:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 25 seconds
{
  "name": "awscli",
  "duration": 10158,
  "function": "aws --endpoint-url http://minikube:30126 s3api head-object --bucket awscli-mint-test-bucket-19989 --key prefix/directory\n",
  "status": "FAIL",
  "error": "An error occurred (404) when calling the HeadObject operation: Not Found"
}

Executed 0 out of 13 tests successfully.
```

**Review by kannappanr — APPROVED** [review]
LGTM

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Pull request merged**